### ISSUE TITLE
Deadlock when recursive resolving scoped services from another thread

### ISSUE THREAD

**Issue by JWilh** [dotnet/runtime#80697]
### Description

Hi,

the service provider locks itself when recursively resolving a scoped service from another thread inside a constructor of a scoped service.
Please check the code example for more details. I'm not able to describe the problem better in plain English :-).

Thanks in advance. 

### Reproduction Steps

```
using Microsoft.Extensions.DependencyInjection;

var provider = new ServiceCollection()
    .AddScoped<TaskRunType>()
    .AddScoped<SecondType>().BuildServiceProvider();

using (var scope = provider.CreateScope())
    scope.ServiceProvider.GetRequiredService<TaskRunType>();

class TaskRunType
{
    public TaskRunType(IServiceProvider provider)
    {
        Task.Run(() =>
        {
            provider.GetRequiredService<SecondType>();
        }).GetAwaiter().GetResult();
    }
}

class SecondType
{ }
```

### Expected behavior

The services should be resolved successfully.

### Actual behavior

Resolving the SecondType ends in a deadlock.

### Regression?

No, doesn't seems to be a regression, at least not lately, this also fails with .net 5.0.

### Known Workarounds

Add a second sync code path to the problematic code section to avoid resolving the second service on another thread.

### Configuration

.Net 7.0.101
Windows 10 21H2
x64

### Other information

I'm aware that this code is very ugly and it's wrong in the first place, it's just the easiest way to trigger this problem.

In our case we normally use a cached value, there is only a very specific scenario where this isn't the case.
We changed the affected code to use async/await and because of that we are now encounter that deadlock.

**Comment by steveharter** [comment]
Thanks for this issue. It looks like it is a duplicate of https://github.com/dotnet/runtime/issues/36458.

Closing as a duplicate.

**Comment by JWilh** [comment]
Hi @steveharter, 
thanks for looking into this.
I saw the referenced thread before, but IMHO these are are different issues.

My example works with singletons, it only deadlocks when used with scoped services and the callstacks are different than from https://github.com/dotnet/runtime/issues/36458:
![image](https://user-images.githubusercontent.com/54053197/215982898-021769d5-8165-4e4a-9c08-c94be33ce78a.png)

**Comment by steveharter** [comment]
>Hi @steveharter,
thanks for looking into this.
I saw the referenced thread before, but IMHO these are are different issues.
Can you please reopen this issue?

Sorry I didn't see that. I was able to repro this on 7.0 but unable to repro on 8.0 preview. Can you please verify? Thanks

**Comment by steveharter** [comment]
> I'm able to reproduce this using .net v8.0.100-preview.1 with the package reference:

Thanks. I was also able to repro it on 8; not sure why I couldn't repro it earlier.

**Comment by JWilh** [comment]
You can close this issue if you like, as I said in the opening post, it's crazy to do it in the first place 😅.

I was just hoping it's an "easy" fix, as it was already fixed for singletons.

**Comment by davidfowl** [comment]
introducing a per scope per type lock would introduce lots of overhead. Today we cheat for singletons and store it on an existing object that was already created per type. For scopes, its less obvious how we'd reduce the overhead.

**Comment by wvpm** [comment]
Instead of creating a deadlock and freezing the application, it could time out.

Also note that the DI guidelines explicitly state _"async/await and Task based service resolution isn't supported. Because C# doesn't support asynchronous constructors, use asynchronous methods after synchronously resolving the service."_
Source: https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines
